src/mx/_impl/mx_javamodules.py

```diff
@@ -521,6 +521,9 @@ def _check_uses(uses, context):
             mx.abort(f"specification of service {use} must use non-binary name of nested class (i.e. replace '$' with '.')", context=context)
     return uses
 
+def _jlink_libraries():
+    return not (mx.get_opts().no_jlinking or mx.env_var_to_bool('NO_JLINKING'))
```

Comment:
The `--no-linking` option is [defined in the Graal repo](https://github.com/oracle/graal/blob/88bc81e7b6404eaeccd8689fce4622eb5718f52a/sdk/mx.sdk/mx_sdk_vm_impl.py#L347) so it doesn't seem right to refer to it here. There's also a bunch of other Graal specific logic being introduced by this PR.
It would be better if you could come up with some Graal-agnostic abstractions that can be added to mx distributions that can then be used by Graal-specific distributions to achieve your goals.

Reply:
Nothing immediate comes to mind but I would suggest something like adding a support for an attribute to distributions that can be set dynamically in [`mx_post_parse_cmd_line`](https://github.com/graalvm/mx/blob/2302f2508bd0048fdc0ab04feac6de63261069e7/src/mx/_impl/mx.py#L2128) for the relevant suites based on the `--no-jlinking` opt.

On the other hand, couldn't the logic you want simply be predicated on whether a jmods dir exists? Or maybe a special `NO_JMODS` value for the [`--jmods-dir` option](https://github.com/graalvm/mx/pull/209) could drive this?

Reply:
> Nothing immediate comes to mind but I would suggest something like adding a support for an attribute to distributions that can be set dynamically in [`mx_post_parse_cmd_line`](https://github.com/graalvm/mx/blob/2302f2508bd0048fdc0ab04feac6de63261069e7/src/mx/_impl/mx.py#L2128) for the relevant suites based on the `--no-jlinking` opt.

OK, thanks.

> On the other hand, couldn't the logic you want simply be predicated on whether a jmods dir exists? Or maybe a special `NO_JMODS` value for the [`--jmods-dir` option](https://github.com/graalvm/mx/pull/209) could drive this?

Yes, I think that would be fine.

Reply:
The updated version now adds some abstraction for custom javac args processing, like we need for Mandrel based on JEP 493 enabled base JDKs. It also uses `--jmods-dir NO_JMODS` as a special value to indicate that JMODs should not be used when compiling java modules.